VideoLocker/src/main/java/org/edx/mobile/view/adapters/CourseOutlineAdapter.java

```diff
@@ -278,85 +274,63 @@ private void updateUIForVideo(int position, View convertView, final ViewHolder v
-                                    logger.debug("Download Button Clicked");
-                                    //notifyDataSetChanged();
-                                    mDownloadListener.download(videoData);
-                                }
-                            });
-                        } else if (ds == DownloadEntry.DownloadedState.DOWNLOADING) {
-                            // may be download in progress
-                            EventBus.getDefault().post(new DownloadEvent(DownloadEvent.DownloadStatus.STARTED));
-                            viewHolder.bulkDownload.setVisibility(View.GONE);
-                            storage.getDownloadProgressByDmid(videoData.dmId, new DataCallback<Integer>(true) {
-                                @Override
-                                public void onResult(Integer result) {
-                                    if (result >= 0 && result < 100) {
-                                        EventBus.getDefault().post(new DownloadEvent(DownloadEvent.DownloadStatus.STARTED));
-                                    } else if (result == 100) {
-                                        EventBus.getDefault().post(new DownloadEvent(DownloadEvent.DownloadStatus.COMPLETED));
-                                    }
-                                }
-
-                                @Override
-                                public void onFail(Exception ex) {
-                                    logger.error(ex);
-                                    viewHolder.bulkDownload.setVisibility(View.VISIBLE);
-                                }
-                            });
-                        } else if (ds == DownloadEntry.DownloadedState.DOWNLOADED) {
-                            // downloaded
-                            viewHolder.bulkDownload.setVisibility(View.GONE);
+                    public void onResult(DownloadEntry.WatchedState result) {
+                        if (result != null && result == DownloadEntry.WatchedState.WATCHED) {
+                            viewHolder.rowType.setIconColorResource(R.color.edx_grayscale_neutral_base);
+                        } else {
+                            viewHolder.rowType.setIconColorResource(R.color.edx_brand_primary_base);
                         }
-
                     }
 
                     @Override
                     public void onFail(Exception ex) {
                         logger.error(ex);
-                        viewHolder.bulkDownload.setVisibility(View.VISIBLE);
                     }
                 });
 
+        if (videoData.isVideoForWebOnly()) {
+            viewHolder.numOfVideoAndDownloadArea.setVisibility(View.GONE);
+        }
+        else {
+            viewHolder.numOfVideoAndDownloadArea.setVisibility(View.VISIBLE);
+            dbStore.getDownloadedStateForVideoId(videoData.videoId,
+                    new DataCallback<DownloadEntry.DownloadedState>(true) {
```

Comment:
Note that these asynchronous can be problematic in a list adapter where the rows are being recycles, since slow responses during queries will result in them being applied to incorrect rows. On the other hand synchronous queries on the main thread (as implemented in `getRowViewForContainer()`) are problematic as well, as they introduce a large performance hit in scrolling and other UI refreshes. As a temporary patch, we could disable row recycling for now, so that we can at least guarantee that we're displaying the correct statuses. For the long term, we might want to preload/synchronize these states with our course data, so that we don't have to query the database for the status of each row before displaying it, which is really inefficient and disk-intensive.